compiler/src/dotty/tools/dotc/transform/sjs/ExplicitJSClasses.scala

```diff
@@ -12,54 +17,728 @@ import core.Types._
+  }
+
+  override def transformInfo(tp: Type, sym: Symbol)(using Context): Type = tp match {
+    case tp @ ClassInfo(_, cls, _, decls, _) if !cls.is(JavaDefined) && isApplicableOwner(cls) =>
+      val innerJSClasses = decls.filter(isJSClass)
+
+      val innerObjectsForAdHocExposed =
+        if (!cls.isStaticOwner) Nil // those already have a module accessor
+        else decls.filter(isExposedModule)
+
+      if (innerJSClasses.isEmpty && innerObjectsForAdHocExposed.isEmpty) {
+        tp
+      } else {
+        def addAnnots(sym: Symbol, symForName: Symbol): Unit = {
+          val jsNameAnnot = symForName.getAnnotation(jsdefn.JSNameAnnot).getOrElse {
+            makeJSNameAnnotation(symForName.defaultJSName)
+          }
+          sym.addAnnotation(jsNameAnnot)
+          sym.addAnnotation(jsdefn.ExposedJSMemberAnnot)
+        }
+
+        val clsIsJSClass = cls.hasAnnotation(jsdefn.JSTypeAnnot)
+
+        val decls1 = decls.cloneScope
+
+        for (innerJSClass <- innerJSClasses) {
+          def addAnnotsIfInJSClass(sym: Symbol): Unit = {
+            if (clsIsJSClass)
+              addAnnots(sym, innerJSClass)
+          }
+
+          val fieldName = jsclassFieldName(innerJSClass.name.asTypeName)
+          val fieldFlags = Synthetic | Artifact
+          val field = newSymbol(cls, fieldName, fieldFlags, defn.AnyRefType, coord = innerJSClass.coord)
+          addAnnotsIfInJSClass(field)
+          decls1.enter(field)
+        }
+
+        // scala-js/scala-js#4086 Create exposed getters for exposed objects in static JS objects
+        for (innerObject <- innerObjectsForAdHocExposed) {
+          assert(clsIsJSClass && cls.is(ModuleClass) && cls.isStatic,
+              i"trying to ad-hoc expose objects in non-JS static object ${cls.fullName}")
+
+          val getterName = jsobjectGetterNameFor(innerObject)
+          val getterFlags = Method | Synthetic | Artifact
+          val getter = newSymbol(cls, getterName, getterFlags, ExprType(defn.AnyRefType), coord = innerObject.coord)
+          addAnnots(getter, innerObject)
+          decls1.enter(getter)
+        }
+
+        tp.derivedClassInfo(decls = decls1)
+      }
+
+    case _ =>
+      tp
+  }
+
+  /** Adjust the `NoInits` flag of Scala traits containing a JS class and of JS traits. */
+  override def transform(ref: SingleDenotation)(using Context): SingleDenotation = {
+    super.transform(ref) match {
+      case ref1: SymDenotation if ref1.is(Trait, butNot = JavaDefined) =>
+        val isJSType = ref1.hasAnnotation(jsdefn.JSTypeAnnot)
+        if (ref1.is(NoInits)) {
+          // If one of the decls is a JS class, there is now some initialization code to create the JS class
+          if (!isJSType && ref1.info.decls.exists(isJSClass))
+            ref1.copySymDenotation(initFlags = ref1.flags &~ NoInits)
+          else
+            ref1
+        } else {
+          // JS traits never have an initializer, no matter what dotc thinks
```

Comment:
Perhaps the choice of words is not appropriate, here.
In a way, adding the `NoInits` flag here precisely makes dotty aware that a JS trait does not have an initializer. It's just that the initial computation of whether or not to put `NoInits` does not add the flag to JS traits, since it is not directly aware of the specificity of JS traits. That initial computation happens in Namer. Maybe I could try to add that directly in Namer, but then I would also have to do something in Scala2Unpickler to fix up things coming from Scala 2 (there is already some logic in there to patch `NoInits` on some well-known Scala 2 traits, because they never have the NoInits flag in the pickles).